ingestion-edge/ingestion_edge/publish.py

```diff
@@ -3,44 +3,135 @@
 # file, you can obtain one at http://mozilla.org/MPL/2.0/.
 
 from .conf import ROUTE_TABLE
+from .conf import QUEUE_PATH
 from datetime import datetime
 from flask import Blueprint, request
-from google.cloud.pubsub import PublisherClient
+from google.api_core.exceptions import Aborted
+from google.api_core.exceptions import Cancelled
+from google.api_core.exceptions import Forbidden
+from google.api_core.exceptions import RetryError
+from google.api_core.exceptions import ServerError
+from google.api_core.exceptions import TooManyRequests
+from google.cloud.pubsub_v1 import PublisherClient
+from google.cloud.pubsub_v1.publisher.exceptions import PublishError
+from google.cloud.pubsub_v1.types import PubsubMessage
+from persistqueue import SQLiteAckQueue
+from persistqueue.exceptions import Empty
 from typing import Tuple
+import json
+
+FLUSH_EMPTY_STATUS = 204
+FLUSH_COMPLETE_STATUS = 200
+FLUSH_INCOMPLETE_STATUS = 504
+# Errors where message should be queued locally after pubsub delivery fails
+# To minimize data loss include errors that may require manual intervention
+TRANSIENT_ERRORS = (
+    # The API thinks the client should retry
+    Aborted,
+    # The API was interrupted when handling the request
+    Cancelled,
+    # The PublisherClient lacks permissions and may require manual intervention
+    Forbidden,
+    # API call ran out of retries
+    RetryError,
+    # API outage or connection issue
+    ServerError,
+    # A python Future timed out
+    TimeoutError,
+    # The API is throttling requests
+    TooManyRequests,
+)
 
 publish = Blueprint('publish', __name__)
 client = PublisherClient()
+q = SQLiteAckQueue(QUEUE_PATH)
```

Comment:
Did we switch to the SQLite queue because it was more stable? Can we add a comment here to give some context?

Reply:
we switched to `SQLiteAckQueue` because `Queue` does not implement acks, which we need.

Reply:
that said, we should use a SQLite queue even if we don't need `ack` because it provides multi-process safe (and threadsafe if we had threads) access to the queue and it's [faster than `Queue`](https://github.com/peter-wangxu/persist-queue/issues/68).

Reply:
also:

> Can we add a comment here to give some context?

absolutely